llvmlite/binding/ffi.py

```diff
@@ -151,20 +152,34 @@ def __call__(self, *args, **kwargs):
             return self._cfn(*args, **kwargs)
 
 
-_lib_name = get_library_name()
-
-
-pkgname = ".".join(__name__.split(".")[0:-1])
-try:
-    _lib_handle = importlib.resources.path(pkgname, _lib_name)
-    lib = ctypes.CDLL(str(_lib_handle.__enter__()))
-    # on windows file handles to the dll file remain open after
-    # loading, therefore we can not exit the context manager
-    # which might delete the file
-except OSError as e:
-    msg = f"""Could not find/load shared object file: {_lib_name}
- Error was: {e}"""
-    raise OSError(msg)
+_lib_load_errors = []
+# llvmlite native code may exist in one of two places, with the first taking
+# priority:
+# 1) In a shared library available as a resource of the llvmlite package.
+#    This may involve unpacking the shared library from an archive.
+# 2) Linked directly into the main binary. Symbols may be resolved from the
+#    main binary by passing None as the argument to ctypes.CDLL.
+for lib_context in (
+        importlib.resources.path(
+            __name__.rpartition(".")[0], get_library_name()),
+        contextlib.nullcontext(None)):
+    try:
+        with lib_context as lib_path:
+            lib = ctypes.CDLL(lib_path and str(lib_path))
```

Comment:
What is the intent of using `and` here. Is this something special about the type of `lib_context?